### PR TITLE
Fix double-sync crash on postgres 9.x

### LINKED ISSUE #1105: TypeError: Cannot read property 'name' of null
First off, thanks for the library! Overall it's been incredibly performant and straightforward to use. I am running into one issue which may be an issue with the library, though it's a bit hard to follow the protocol for me so my usage may be off as well.

I'm running into an uncaught error during the `parseComplete` handler in client.js (https://github.com/brianc/node-postgres/blob/v6.0.1/lib/client.js#L131).

I am doing a direct `connect` against the pool and then utilizing pg-query-stream

``` javascript
pool.connect().then(function(client){
  stream = client.query(new QueryStream(query, null));

  stream.on('error', function(streamError){
    stream.close(function(closeError){ 
      // What to do with a close error? At this point stream is already hosed :shrug:
      client.release(streamError);
    }
  });

  stream.on('close', function(){
    client.release();
  }); 
});
```

At some point during my app's life cycle, I get an uncaught error

```
TypeError: Cannot read property 'name' of null
    at null.<anonymous> (/var/app/current/node_modules/pg/lib/client.js:139:26)
    at emit (events.js:107:17)
    at Socket.<anonymous> (/var/app/current/node_modules/pg/lib/connection.js:121:12)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:540:20)
```

So obviously `activeQuery` is becoming null, and I am trying to narrow down why I would receive a `parseComplete` event when the active query is null.

According to my logs, the stream is _not_ emitting an error, and seems to be closing normally. Therefore it seems like the connection client is getting a `readyForQuery` or `end` event, and then right after that getting the `parseComplete` event. Any idea why this would be happening, or see any issues with my usage? Thanks for any help you can give! I'll keep looking into it as well.

### PR DESCRIPTION
**Pull request by brianc**
An attempt to fix #1105

It looks like in rare, random occasions postgres 9.x is sending both `ErrorResponse` __and__ `CommandComplete` when a query times out.  In an attempt to fix this, I have changed the query to only send a `sync` on an error message if it hasn't already sent a sync.  Judging from the docs this should never happen, and indeed on newer versions of postgres the bug is not reproducible...but lots of folks still run postgres 9.x in production.  I'm not super fond of this fix as it feels like it's working around a possible race condition in postgres 9.x so I'd love some additional eyes & ideas on this one.  Also, want to run the test suite and make sure everything passes in the whole test matrix on travis.

One option is to put this behind some environment variable like `PGSKIPDOUBLESYNC` or something and allow people to opt in to this change.  I'm worried about a (according to the docs, not possible) circumstance where we receive two messages requiring a sync and we only send one sync back.

### CODE REVIEW

**Comment by brianc** [comment]
@charmander @sehrope any thoughts here?  All tests are passing which is 👍 but it feels like I'm working around an issue or possible race condition w/in older versions of postgres itself?

**Comment by charmander** [comment]
@brianc Did you build a tool that could reproduce this pretty reliably in order to narrow it down to 9.x, or does the evidence come from somewhere else? Maybe we can narrow it down to a specific version and find the change in PostgreSQL to confirm the fix.

**Comment by sehrope** [comment]
I'm able to reproduce this error (without the fix) on 9.3 through 10 (not just 9.x). I can't reproduce it on v11, v12, or v13 after a bunch of repeated runs. It also seems less frequent on v10 vs earlier versions so I'm guessing it is some kind of server race condition to cause the difference. Hoping to have some more time to look into it later this week.

**Comment by sehrope** [comment]
Oh and can confirm that @brianc's fix removes the issue on those older versions. At least in my repeated testing.

Still feels weird working around a potential server FEBE bug. If there's truly a protocol break in older versions might make more sense to scrap the entire connection as who knows what else could be out of sync. Thrashing connection pools is never fun but I worry either this or the fix would lead to an error in some other location too.

File: packages/pg/test/integration/gh-issues/1105-tests.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const pg = require('../../../lib')
+const helper = require('../test-helper')
+const suite = new helper.Suite()
+
+suite.testAsync('timeout causing query crashes', async () => {
+  const client = new helper.Client()
+  await client.connect()
+  await client.query('CREATE TEMP TABLE foobar( name TEXT NOT NULL, id SERIAL)')
+  client.query('BEGIN')
```
Comment (sehrope) [inline packages/pg/test/integration/gh-issues/1105-tests.js:9]: s/client/await client/

Comment (brianc) [reply packages/pg/test/integration/gh-issues/1105-tests.js:9]: ah yeah that's a hold over when i thought it might have been related to the query queue (which is a horrible unfortunate API design mistake I made 10 years ago).  So I initially wrote the test kinda abusing that feature thinking it was something I'd done.  I just added await back here, thanks!

**Comment by brianc** [comment]
> Still feels weird working around a potential server FEBE bug. If there's truly a protocol break in older versions might make more sense to scrap the entire connection as who knows what else could be out of sync.

Yeah I'm asking in #postgres on IRC to see if anyone on the postgres development team knows more about this.  This isn't a bad idea, i'll look at doing this as well.

**Comment by brianc** [comment]
Okay [here](https://travis-ci.org/github/brianc/node-postgres/builds/734004743) is the matrix of test runs which fail without the patch applied.

**Comment by brianc** [comment]
> If there's truly a protocol break in older versions might make more sense to scrap the entire connection as who knows what else could be out of sync.

It's not trivial to trash a client and error out any pending and in-flight queries and stop processing other incoming messages w/ the current architecture.  I think this might risk introducing more places things could go wrong and unexpected errors or behaviors, but I'll fiddle with it.  I am thinking one "safer" (less invasive) fix could be to have an environment variable like `PGSKIPDOUBLESYNC=true` or something & folks can opt-in to the fix?

**Comment by brianc** [comment]
okay...so w/ some help from @sehrope and a postgres maintainer from IRC I've changed the approach somewhat.  When discussing the maintainer (I believe their github name is @rhodiumtoad) suggested to pipeline the `sync` message at the end of the other pipelined messages.  Sending `sync` immediately after `execute` allows the backend to process all the messages and only respond with _either_ `CommandComplete` or `ErrorResponse`.  This fixes the issue in the test w/o requiring any hacks or discarding messages.  It also speeds up some of my crude benchmarks by over 10% which is a super awesome bonus side effect.

File: packages/pg/lib/query.js [hunk 2]
```diff
@@ -141,6 +130,16 @@ class Query extends EventEmitter {
     this.emit('error', err)
   }
 
+  handleReadyForQuery(con) {
+    if (this._canceledDueToError) {
+      return this.handleError(this._canceledDueToError, con)
```
[the remark this answers is not in the document]
Comment (junaid1460) [reply packages/pg/lib/query.js:135]: cancelled

[the remark this answers is not in the document]
Comment (sehrope) [reply packages/pg/lib/query.js:135]: It's not a typo. In American English canceled with a single "L" is the preferred one.
